### PR TITLE
Suport Ascend NPU

### PR DESCRIPTION
**Pull request by hipudding**
Ascend NPU is a series of AI processors.
[hiascend](https://www.hiascend.com/).

This commit make InputArray and OutputArray support a new Mat structor called NpuMat, which has a same interface with Mat and GpuMat. And also, python bindings generator support NpuMat for python interface.

This PR is only for type declaration and code generation. For implementation, please refer to [opencv_contrib#3552](https://github.com/opencv/opencv_contrib/pull/3552)

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- N/A There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by asmorkalov** [comment]
@hipudding Thanks a lot for the contribution!
General comments on the integration:
- I propose to add Ascend to constant names like NPU_MAT. NPU is too generic and there are several NPU vendors.
- Most of pages related to Ascend are in Chinese. Please add setup guide to tutorials in English. Also it'll be great to have future proof solution, because it's hard to navigate through pages in Chinese.
- NPU usage model and restrictions are not clean: what about thread safety, relevant image sizes, etc. Please add some documentation/tutorial about it.
- It'll be great to have Dockerfile with build environment for CI/Experiments.

**Comment by hipudding** [comment]
> @hipudding Thanks a lot for the contribution!
> 
> General comments on the integration:
> 
> - I propose to add Ascend to constant names like NPU_MAT. NPU is too generic and there are several NPU vendors.
> 
> - Most of pages related to Ascend are in Chinese. Please add setup guide to tutorials in English. Also it'll be great to have future proof solution, because it's hard to navigate through pages in Chinese.
> 
> - NPU usage model and restrictions are not clean: what about thread safety, relevant image sizes, etc. Please add some documentation/tutorial about it.
> 
> - It'll be great to have Dockerfile with build environment for CI/Experiments.

Sure. Will add them ASAP.

File: modules/core/include/opencv2/core/mat.hpp [hunk 1]
```diff
@@ -183,7 +183,8 @@ class CV_EXPORTS _InputArray
 #if OPENCV_ABI_COMPATIBILITY < 500
         STD_ARRAY         =14 << KIND_SHIFT,  //!< removed: https://github.com/opencv/opencv/issues/18897
 #endif
-        STD_ARRAY_MAT     =15 << KIND_SHIFT
+        STD_ARRAY_MAT     =15 << KIND_SHIFT,
+        NPU_MAT           =16 << KIND_SHIFT
```
Comment (opencv-alalek) [inline modules/core/include/opencv2/core/mat.hpp:187]: For such customization it is better to use approach described here: #20869

Comment (hipudding) [reply modules/core/include/opencv2/core/mat.hpp:187]: sure. Will fix it in next commit.

**Comment by hipudding** [comment]
> * Most of pages related to Ascend are in Chinese. Please add setup guide to tutorials in English. Also it'll be great to 
have future proof solution, because it's hard to navigate through pages in Chinese.
Here's english version website of Ascend(https://www.hiascend.com/en/document), Please have a try and see whether it can provide enough informations.

> * It'll be great to have Dockerfile with build environment for CI/Experiments.
Could you give me the visitor's IP address and public key(security information, please send me by email: huafengchun@gmail.com). I will setup a CI VM for opencv,  You will be the managers of this machine. And also, I will add a Dockerfile in this PR for running testcases.

**Comment by hipudding** [comment]
@asmorkalov @opencv-alalek  I  found CI is failed due to added/removed symbols [here](https://pullrequest.opencv.org/buildbot/builders/precommit_linux64/builds/104605/steps/Compare%20ABI%20dumps/logs/report-html#Constant_Problems_Low). But I found these symbols is not relate to this PR. Could you give me some suggestions to pass it? Thanks.

**Comment by MengqingCao** [comment]
> * NPU usage model and restrictions are not clean: what about thread safety, relevant image sizes, etc. Please add some documentation/tutorial about it.

Thanks for your suggestion, the tutorial of Ascend NPU image processing is already available and included in “_opencv_contrib/modules/cannops/tutorials/ascend_npu_image_processing.markdown_”.

File: modules/core/include/opencv2/core/types_c.h [hunk 2]
```diff
@@ -196,7 +196,8 @@ enum {
  CV_OpenCLApiCallError=        -220, /**< OpenCL API call error */
  CV_OpenCLDoubleNotSupported=  -221,
  CV_OpenCLInitError=           -222, /**< OpenCL initialization error */
- CV_OpenCLNoAMDBlasFft=        -223
+ CV_OpenCLNoAMDBlasFft=        -223,
+ CV_AscendApiCallError =       -224  /**< AscendCL API call error */
```
[the remark this answers is not in the document]
Comment (hipudding) [reply modules/core/include/opencv2/core/types_c.h:200]: Thanks for your reivew. Will fix them in next commit.

[the remark this answers is not in the document]
Comment (hipudding) [reply modules/core/include/opencv2/core/types_c.h]: Fixed.

**Pull request closed without merging**